tools/2025-01-gitops-technical-design.md

```diff
@@ -0,0 +1,435 @@
+
+The controller is set up to watch `GitRepository` resources in the cluster:
+```go
+func (r *FluxController) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &radappiov1alpha3.DeploymentTemplate{}, deploymentTemplateRepositoryField, deploymentTemplateRepositoryIndexer); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&sourcev1.GitRepository{}, builder.WithPredicates(GitRepositoryRevisionChangePredicate{})).
+		Complete(r)
+}
+```
+
+When a `GitRepository` resource is created or updated, the controller will get the object and read the contents of the Git repository:
+```go 
+	// Get the GitRepository object from the cluster
+	var repository sourcev1.GitRepository
+	if err := r.Get(ctx, req.NamespacedName, &repository); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+  ...
+
+  // Fetch the artifact from the Source Controller
+	logger.Info("Fetching artifact", "url", artifact.URL)
+	if err := r.ArchiveFetcher.Fetch(artifact.URL, artifact.Digest, tmpDir); err != nil {
+		logger.Error(err, "unable to fetch artifact")
+		return ctrl.Result{}, err
+	}
+```
+
+#### Parsing the `radius-config.yaml` File
+
+The controller will look for a `radius-config.yaml` file in the Git repository. If the file is not present, then the controller will skip the repository. If the file is present, then the controller will read the file and extract the Bicep compilation instructions:
+```go
+  // Read the radius-config.yaml file
+  configFile := filepath.Join(tmpDir, "radius-config.yaml")
+  config, err := ioutil.ReadFile(configFile)
+  if err != nil {
+    logger.Error(err, "unable to read radius-config.yaml file")
+    return ctrl.Result{}, err
+  }
+
+  // Parse the radius-config.yaml file
+  var radiusConfig RadiusConfig
+  if err := yaml.Unmarshal(config, &radiusConfig); err != nil {
+    logger.Error(err, "unable to parse radius-config.yaml file")
+    return ctrl.Result{}, err
+  }
+```
+
+The `radius-config.yaml` file will have the following structure:
+> Note: The structure of the `radius-config.yaml` file is subject to change. We should discuss this in the design meeting. I considered adapting the existing `.rad/config.yaml` file format, but there were too many differences between the two formats, so I decided to create a new format specifically for the Radius Flux Controller.
+```yaml
+config:
+  bicep:
+    - file: main.bicep
+      parameters: main.bicepparam
+      namespace: default
+    - file: secondary.bicep
+  providers:
+    - radius:
+        scope: /planes/radius/local/resourcegroups/default
+    - aws:
+        scope: /planes/aws/aws/accounts/000000000/regions/us-west-2
+    - azure:
+        scope: 	/planes/azure/azurecloud/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/my-group
```

Comment:
What's the motivation for adding providers here, given that they are already expected to be configured in env.bicep?

Reply:
good point, removed